lib/code-path-analysis/code-path-analyzer.js

```diff
@@ -0,0 +1,350 @@
+        case "ArrowFunctionExpression":
+            shiftToHeadSeg(analyzer, node);
+            analyzer.codePathStack.pop();
+            break;
+
+        case "IfStatement":
+        case "ConditionalExpression":
+            if (node.alternate == null) {
+                state.forkInheritPath();
+            }
+            state.popSegContext();
+            break;
+
+        case "LogicalExpression":
+            state.popSegContext();
+            break;
+
+        case "SwitchStatement":
+            state.popSwitchContext();
+            break;
+
+        case "SwitchCase":
+            // This is the same as the process at 1st `consequent` node in `preprocess` function.
+            // Must do if this `consequent` is empty.
+            if (node.consequent.length === 0) {
+                state.makeSwitchCaseBody(
+                    true,
+                    isDefaultNode(node),
+                    lastOf(node.parent.cases) === node);
+            }
+            break;
+
+        case "TryStatement":
+            state.popTryContext();
+            break;
+
+        case "BreakStatement":
+            shiftToHeadSeg(analyzer, node);
+            state.makeBreak(node.label && node.label.name);
+            break;
+
+        case "ContinueStatement":
+            shiftToHeadSeg(analyzer, node);
+            state.makeContinue(node.label && node.label.name);
+            break;
+
+        case "ReturnStatement":
+            shiftToHeadSeg(analyzer, node);
+            state.makeReturn();
+            break;
+
+        case "ThrowStatement":
+            shiftToHeadSeg(analyzer, node);
+            state.makeThrow();
+            break;
+
+        case "WhileStatement":
+        case "DoWhileStatement":
+            state.popLoopContext();
+            break;
+
+        default:
+            break;
+    }
+}
+
+//------------------------------------------------------------------------------
+// Public Interface
+//------------------------------------------------------------------------------
+
+function CodePathAnalyzer(emitter) {
+    this.emitter = emitter;
+    this.codePathStack = [];
+}
+
+CodePathAnalyzer.prototype = {
+    constructor: CodePathAnalyzer,
+
+    get codePath() {
+        return lastOf(this.codePathStack);
+    },
+
+    enterNode: function enterNode(node) {
+        var codePath = this.codePath;
+
+        // Updates code path.
+        processCodePathToEnter(this, node);
+
+        // Emits CodePath/CodePathSegment events.
+        if (codePath !== this.codePath) {
+            codePath = this.codePath;
+            debug.dump("*)", "onCodePath");
+            this.emitter.emit("onCodePath", codePath, node);
+        }
+        shiftToHeadSeg(this, node);
+
+        var state = CodePath.getState(codePath);
+        if (debug.enabled) {
+            state.currentSeg.internal.nodes.push(node);
+        }
+
+        // Emits node events.
+        debug.dumpState(node.type, state);
+        this.emitter.emit(
```

Comment:
Oh, Yes my bad.

I will add `context.getCodePath()` similar to `context.getScope()`, instead.